docs/proposals/24-shoot-oidc-issuer.md

```diff
@@ -0,0 +1,128 @@
+
+This GEP proposes a mechanism that will allow users to expose such information through Gardener managed domain.
+
+## Motivation
+
+By default Gardener creates clusters which expose such documents directly through the `kube-apiserver`. This has several drawbacks:
+- The `kube-apiserver` is protected and unless `--annonymous-auth` is enabled then the endpoint requires authn.
+- The certificate used to serve the documents is not signed by a trusted CA.
+
+Gardener provides means for users to tweak configuration and overcome these drawbacks by using the following [shoot configurations](https://github.com/gardener/gardener/blob/580324da9af4ec47955d9e216569d09053c5d008/example/90-shoot.yaml#L201-L204) which directly interact with the [--service-account-issuer](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-apiserver/) flag of the `kube-apiserver`. Once the issuer of a shoot is changed the user can take over that domain and expose the OIDC metadata documents to the public. This poses another challenge since it requires additional involvement. Some users use istio/nginx igress to do so, others make use of projects like [service-account-issuer-discovery](https://github.com/gardener/service-account-issuer-discovery), or combination of both.
+
+A simpler approach will be for Gardener to be able to take ownership of a shoot's issuer in a central place which can serve all shoots OIDC discovery documents. This will facilitate easier integrations and even allow clusters residing in fenced environments, i.e., those protected by a firewall, to share their OIDC documents. This approach is in sync with the concept of a managed service and is what other Kubernetes offerings as GKE, AKS and EKS already provide.
+
+### Goals
+ - Stay backwards compatible with the current configuration options. Existing integrations should continue to work.
+ - Provide a secure managed solution for serving the OIDC discovery documents of a shoot cluster.
+ - Improve integration capabilities for clusters that reside in a fenced environment.
+### Non-Goals
+ - Establish trust relations between clusters. This GEP is about simplifying the prerequisites for doing that.
+
+## Proposal
+
+This proposal consists of three parts that together will achieve the mentioned goals.
+
+### Static validation
+
+Static validation rule will be introduced enforcing that `shoot.Spec.Kubernetes.KubeAPIServer.ServiceAccountConfig.Issuer` is not set when the shoot is annotated with `gardener.cloud/managed-issuer: true`. The rule will also assure that once present the annotation cannot be removed from a shoot.
```

Comment:
> The rule will also assure that once present the annotation cannot be removed from a shoot.

Why is this restriction necessary?

> the shoot is annotated with `gardener.cloud/managed-issuer: true`

For the time the feature is being alpha/beta tested it sounds nice to be set via annotation, but once the feature graduates to stable can it be set via field in the shoot API? Do you have any disadvantages in mind against API field?

Reply:
> > The rule will also assure that once present the annotation cannot be removed from a shoot.
> 
> Why is this restriction necessary?
> 
> > the shoot is annotated with `gardener.cloud/managed-issuer: true`
> 
> For the time the feature is being alpha/beta tested it sounds nice to be set via annotation, but once the feature graduates to stable can it be set via field in the shoot API? Do you have any disadvantages in mind against API field?

The restriction is not 100% necessary. Here are my thoughts about why I would like to keep the update only in one direction:
- Once enabled this issuer URL will be used so that trust is established to that cluster. Reverting this configuration on accident can lead to broken integrations and downtimes.
- It is easier for users that Gardener manages this on their behalf. I do not think that there is a big case in favour of someone wanting to manage this on their own. Once enabled there is no reason to go back.
- Providers like EKS and GCP have this always on. AKS provides a one way switch to enable this feature (similar to what is proposed here) and you cannot go back once enabled.
- The handling of the logic is simpler this way.

I thought about extending the shoot API, but here are my reasons against doing that:
- The shoot spec is already big enough and I did not find a suitable place for this configuration. The `kubeAPIServer` config is probably not the best place for introducing that since it is also used by the `Garden` specification. It will also introduce coupling between the new field and the `serviceAccountConfig.issuer` (this is already the case with the annotation, but the annotation can be removed later).
- I thought about graduation which in my mind would mean enabling this feature for all new clusters + clusters which do not make use of the `serviceAccountConfig.issuer`. We can eventually even deprecate `serviceAccountConfig.issuer` + `serviceAccountConfig.acceptedIssuers` if the currently proposed feature proves adoption and is what users prefer. Eventually we can remove the annotation as well.

Reply:
> Reverting this configuration on accident can lead to broken integrations and downtimes.

Hibernating the shoot or adding taints to the nodes evicting all workload pods also can lead to a downtime. I guess there are many more options how to break up via shoot spec, so I would not accept this as argument. 

Some users might decide to keep their OIDC discovery documents private, even though the content can be considered as public information, that is their choice and it would be nice if Gardener implement this feature, it could be considered as differentiator compared to GKE/AKS/EKS. 
Codewise, I still think it would not be anything complex, it just needs to delete the configmap with the OIDC documents and use the default issuer instead of the publicly accessible one. The consequence is that existing tokens will be immediately invalidated and will need rotation, but this is the case also when the managed OIDC issuer is enabled, so the code handling the change will be there.

> and I did not find a suitable place for this configuration

I was think about new boolean field `shoot.spec.kubernetes.kubeAPIServer.serviceAccountConfig.managedIssuer`, it will have the same semantic as the annotation and will require the same validation. The advantage of API field is that it will be properly documented and visible from the api spec reference. Gardener already has so many annotations for different purposes that they are becoming unusable from the end users (shoot owners or Gardener operators).

Reply:
> Hibernating the shoot or adding taints to the nodes evicting all workload pods also can lead to a downtime. I guess there are many more options how to break up via shoot spec, so I would not accept this as argument.

Yes, the existence of such APIs does not mean that every newly introduced feature should have an on/off button. We can always loosen the check if there is the need for this, but as of now I am not aware of such use case.

> Some users might decide to keep their OIDC discovery documents private, even though the content can be considered as public information

Why? Is there such use case? We also do not give them access to the private key that signs the service account tokens. See https://github.com/gardener/gardener/pull/7242 as it was removed with this PR. This shows that not every knob brings value to the users. Enabling the feature by default or the deprecation of the other two fields is not in the scope of this GEP. This proposal should not restrict our future options. 

> I was think about new boolean field shoot.spec.kubernetes.kubeAPIServer.serviceAccountConfig.managedIssuer

I do not think that this makes sense in the context of the `virtual-kube-apiserver`. Or do you mean that we only introduce this field in the shoot spec?

From my experience even having the two fields `serviceAccountConfig.issuer` and `serviceAccountConfig.acceptedIssuer` has proven confusing enough for users. Adding a 3rd field is also possible if that is the consensus.

> The advantage of API field is that it will be properly documented and visible from the api spec reference

I agree, although there are a lot of examples of annotations bringing functionality.

Reply:
Can't we assume that Gardener takes over management when `shoot.Spec.Kubernetes.KubeAPIServer.ServiceAccountConfig.Issuer` is unset? (can be combined with Gardenlet feature gate for sake of feature graduation).

> We can eventually even deprecate serviceAccountConfig.issuer + serviceAccountConfig.acceptedIssuers

Can you explain why `serviceAccountConfig.acceptedIssuers` can be deprecated? Isn't that still required to configure trust?

Reply:
> Can't we assume that Gardener takes over management when shoot.Spec.Kubernetes.KubeAPIServer.ServiceAccountConfig.Issuer is unset? (can be combined with Gardenlet feature gate for sake of feature graduation).

This can be a breaking change for users that first create a shoot cluster and then modify its `shoot.Spec.Kubernetes.KubeAPIServer.ServiceAccountConfig.Issuer` as part of some workflow/automatization process. Another problem with this approach is that validation will be harder since we will not have an indication that the shoot has this feature enabled. Do you mean that the gardenlet can set the shoot annotation in case the issuer is empty? I would not mind such approach, but I think that we should collect evidence that existing integrations will not break because of it.

> We can eventually even deprecate serviceAccountConfig.issuer + serviceAccountConfig.acceptedIssuers
Can you explain why serviceAccountConfig.acceptedIssuers can be deprecated? Isn't that still required to configure trust?

The existence of the `acceptedIssuers` field is because once the user provides a new value for `issuer` existing service account tokens will not be accepted by the `kube-apiserver`. It was meant to allow changing the `issuer` in a non disruptive manner. See the example below:

Old config:
```yaml
issuer: https://abc.example
```

New config:
```yaml
issuer: https://new-issuer.example
acceptedIssuers: 
- https://abc.example
```

If we have fully managed setup for all shoot cluster and all of them have the managed issuer feature ON, then we do not need this `acceptedIssuer` field exposed since users would not have the need to interact with it. If for some reason these values need to be changed, Gardener internals can take care without exposing such fields to the user.

Reply:
We seem a bit stuck in this conversion. I'm not sure if we need this kind of validation, it's already allowed to change the `issuer` today.
What do you think of the following?
- Leave out the planned validation
- Introduce feature with documentation first (optional)
- Default fields if unset later